### ISSUE TITLE
 type 'List<Comparable<dynamic>>' is not a subtype of type 'Iterable<Pair>' of 'iterable'

### ISSUE THREAD

**Issue by spkingr** [luanpotter/flame#13]
When I make a simple game, I got the error again and again, how to solve it, thanks!

```
Another exception was thrown: type 'List<Comparable<dynamic>>' is not a subtype of type 'Iterable<Pair>' of 'iterable'
```

And the stack:

```
I/flutter (20726): When the exception was thrown, this was the stack:
I/flutter (20726): #0      List.setRange (dart:core/runtime/libarray.dart)
I/flutter (20726): #1      sort (package:box2d/src/buffer_utils.dart:151:8)
I/flutter (20726): #2      DefaultBroadPhaseBuffer.updatePairs (file:///D:/Software/Google/Flutter/SDK/flutter/.pub-cache/hosted/pub.flutter-io.cn/box2d-0.4.0/lib/src/collision/broadphase/default_broadphase_buffer.dart:140:17)
I/flutter (20726): #3      ContactManager.findNewContacts (file:///D:/Software/Google/Flutter/SDK/flutter/.pub-cache/hosted/pub.flutter-io.cn/box2d-0.4.0/lib/src/dynamics/contact_manager.dart:165:16)
I/flutter (20726): #4      World.stepDt (file:///D:/Software/Google/Flutter/SDK/flutter/.pub-cache/hosted/pub.flutter-io.cn/box2d-0.4.0/lib/src/dynamics/world.dart:566:23)
I/flutter (20726): #5      Box2DComponent.update (package:project_simple_game/flame/box2d/box2d_component.dart:39:11)
I/flutter (20726): #6      GameWorld.update (package:project_simple_game/game_world.dart:26:11)
I/flutter (20726): #7      GameRoot.update (package:project_simple_game/game_root.dart:33:21)
I/flutter (20726): #8      _GameRenderBox._update (package:project_simple_game/flame/game.dart:119:10)
I/flutter (20726): #9      _GameRenderBox._tick (package:project_simple_game/flame/game.dart:112:5)
I/flutter (20726): #10     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:990:15)
I/flutter (20726): #11     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleBeginFrame.<anonymous closure> (package:flutter/src/scheduler/binding.dart:906:11)
I/flutter (20726): #12     __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.forEach (dart:collection/runtime/libcompact_hash.dart:363:8)
I/flutter (20726): #13     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleBeginFrame (package:flutter/src/scheduler/binding.dart:904:17)
I/flutter (20726): #14     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.scheduleWarmUpFrame.<anonymous closure> (package:flutter/src/scheduler/binding.dart:747:7)
I/flutter (20726): #16     _Timer._runTimers (dart:isolate/runtime/libtimer_impl.dart:382:19)
I/flutter (20726): #17     _Timer._handleMessage (dart:isolate/runtime/libtimer_impl.dart:416:5)
I/flutter (20726): #18     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:165:12)
I/flutter (20726): (elided one frame from package dart:async)
I/flutter (20726): 
I/flutter (20726): 
I/flutter (20726): This exception was thrown in the context of a scheduler callback. When the scheduler callback was
I/flutter (20726): _registered_ (as opposed to when the exception was thrown), this was the stack:
I/flutter (20726): #0      new _FrameCallbackEntry.<anonymous closure> (package:flutter/src/scheduler/binding.dart:109:33)
I/flutter (20726): #1      new _FrameCallbackEntry (package:flutter/src/scheduler/binding.dart:112:6)
I/flutter (20726): #2      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.scheduleFrameCallback (package:flutter/src/scheduler/binding.dart:426:53)
I/flutter (20726): #3      _GameRenderBox._scheduleTick (package:project_simple_game/flame/game.dart:102:50)
I/flutter (20726): #4      _GameRenderBox._tick (package:project_simple_game/flame/game.dart:111:5)
I/flutter (20726): #5      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:990:15)
I/flutter (20726): #6      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleBeginFrame.<anonymous closure> (package:flutter/src/scheduler/binding.dart:906:11)
I/flutter (20726): #7      __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.forEach (dart:collection/runtime/libcompact_hash.dart:363:8)
I/flutter (20726): #8      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleBeginFrame (package:flutter/src/scheduler/binding.dart:904:17)
I/flutter (20726): #9      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleBeginFrame (package:flutter/src/scheduler/binding.dart:834:5)
I/flutter (20726): #10     _invoke1 (dart:ui/hooks.dart:134:13)
I/flutter (20726): #11     _beginFrame (dart:ui/hooks.dart:105:3)
```

**Comment by luanpotter** [comment]
Could you please provide your code relevant to the error? I can only guess that you created a world (Box2DComponent) and added inside Components that are not bodies (BodyComponent). All components added directly inside a Box2DComponent must be BodyComponents.

**Comment by spkingr** [comment]
Here is my test code:

```
class GameRoot extends Game{
  GameRoot(){
    this._gameWorld = GameWorld();
    this._gameWorld.initializeWorld();
    this._initAssets().then((_) => this._isLoaded = true);
  }

  GameWorld _gameWorld;
  bool _isLoaded = false;
  Size _dimensions;

  Future<Null> _initAssets() async{
    this._dimensions = await Flame.util.initialDimensions();
    await Flame.audio.load('supersoyboy.ogg');
    await Flame.audio.load('dryfire.mp3');
    Flame.audio.loop('supersoyboy.ogg');
  }

  @override
  void render(Canvas canvas) {
    this._gameWorld.render(canvas);
  }

  @override
  void update(double t){
    this._gameWorld.update(t);
  }

  @override
  void resize(Size size){
    this._gameWorld.resize(size);
  }
}
```

And another file:

```
class GameWorld extends Box2DComponent{
  GameWorld():super();

  SuperBoyComponent player;

  @override
  void initializeWorld() {
    //this.add(BackgroundComponent(this.viewport));
    //this.add(GroundComponent(this));
    this.player = SuperBoyComponent(this);
    this.add(this.player);
  }

  @override
  void update(double t) {
    try{
      super.update(t); //This line triggers the error!
    }catch(e){
      print("Error:------------------$e");
    }
    this.cameraFollow(this.player, horizontal: 0.5);
  }
}

class SuperBoyComponent extends BodyComponent{
  static final double playerHeight = 48.0;
  static final double playerWidth = 40.0;

  SuperBoyComponent(Box2DComponent box) : super(box){
    this._loadImages().then((_) => this._isLoading = false);
    this._createBody();
  }

  bool _isLoading = true;
  final images = Map<String, Image>();

  bool _isIdle = true;
  bool _isRun = false;
  bool _isJump = false;
  bool _isSlide = false;

  @override
  void update(double t) {
    this._isIdle = this.body.linearVelocity.x == 0.0;
    this._isRun = this.body.linearVelocity.x != 0.0;
    this._isJump = this.body.getContactList() == null;
  }

  @override
  void renderPolygon(Canvas canvas, List<Offset> points){
    if(this._isLoading){
      return;
    }

    Image image = this._isJump ? this.images["jump_1"] : (this._isIdle ? this.images["run_1"] : (this._isRun ? this.images["run_1"] : this.images["slide_1"]));
    final rect = Rect.fromLTWH(- SuperBoyComponent.playerWidth * 0.5, - SuperBoyComponent.playerHeight * 0.5, SuperBoyComponent.playerWidth, SuperBoyComponent.playerHeight);
    paintImage(canvas: canvas, image: image, rect: rect, fit: BoxFit.contain, flipHorizontally: this.body.linearVelocity.x < 0.0);
  }

  Future<Null> _loadImages() async{
    /// run_1 run_15
    /// jump_1 jump 15
    /// slide_1
    Image image;
    for(int i = 1; i <= 15; i ++){
      image = await Flame.images.load("run_$i.png");
      this.images["run_$i"] = image;
      image = await Flame.images.load("jump_$i.png");
      this.images["jump_$i"] = image;
    }
    image = await Flame.images.load("slide_1.png");
    this.images["slide_1"] = image;
  }

  void _createBody(){
    final shape = PolygonShape()
      ..setAsBoxXY(SuperBoyComponent.playerWidth, SuperBoyComponent.playerHeight);
    shape.centroid.x = SuperBoyComponent.playerWidth * 0.5;
    shape.centroid.y = SuperBoyComponent.playerHeight * 0.5;

    final fixtureDef = FixtureDef()
      ..shape = shape
      ..restitution = 0.0
      ..density = 0.05
      ..friction = 0.2;

    final bodyDef = BodyDef()
      ..linearVelocity = Vector2.zero()
      ..position = Vector2(0.0, 15.0)
      ..bullet = true
      ..type = BodyType.DYNAMIC;

    this.body = this.world.createBody(bodyDef)
      ..createFixtureFromFixtureDef(fixtureDef)
      ..setFixedRotation(true);
  }
}
```

Thanks very much for help!

**Comment by feroult** [comment]
Hi @spkingr, thanks for your code.
I was able reproduce the problem and I'm trying to fix it.

**Comment by feroult** [comment]
It was a bug in Box2D Dart port. Since the original maintainer was not merging this [PR](https://github.com/google/box2d.dart/pull/11) from Feb, we decided to fork the project and create our own package.

You'll need to change any reference to `box2d` package or sub-packages to add the `_flame` suffix:

```
import 'package:box2d_flame/box2d.dart';
```

Just wait @luanpotter to release next Flame version and you should be ok with your code.

Best

**Comment by luanpotter** [comment]
@spkingr the new version was released (0.9.1). Could you verify if the problem is solved? Thanks!

**Comment by spkingr** [comment]
It works now! Happy to go on flutter game. :) @luanpotter  @feroult
